File: practice/src/main/java/com/decade/practice/usecases/ChatOperations.java
package com.decade.practice.usecases;

import com.decade.practice.models.domain.ChatSnapshot;
import com.decade.practice.models.domain.embeddable.ChatIdentifier;
import com.decade.practice.models.domain.entity.Chat;
import com.decade.practice.models.domain.entity.User;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.UUID;

public interface ChatOperations {
        int CONVERSATION_LIMIT = 20;

        void createChat(ChatIdentifier identifier);

        Chat getOrCreateChat(ChatIdentifier identifier) throws NoSuchElementException;

        default Chat getOrCreateChat(UUID owner, UUID partner) throws NoSuchElementException {
                return getOrCreateChat(ChatIdentifier.from(owner, partner));
        }

        List<Chat> listChat(User owner, Integer version, Chat offset, int limit);

        default List<Chat> listChat(User owner, Integer version, Chat offset) {
                return listChat(owner, version, offset, CONVERSATION_LIMIT);
        }

        default List<Chat> listChat(User owner, Integer version) {
                return listChat(owner, version, null, CONVERSATION_LIMIT);
        }

        default List<Chat> listChat(User owner) {
                return listChat(owner, null, null, CONVERSATION_LIMIT);
        }

        ChatSnapshot getSnapshot(Chat chat, User owner, int atVersion);
}

File: practice/src/main/java/com/decade/practice/database/ChatOperations.kt
package com.decade.practice.database

import com.decade.practice.model.embeddable.ChatIdentifier
import com.decade.practice.model.entity.Chat
import com.decade.practice.model.entity.User
import com.decade.practice.model.local.ChatSnapshot
import com.decade.practice.util.toIdentifier
import java.util.*

private const val CONVERSATION_LIMIT: Int = 20

interface ChatOperations {

    @Throws(NoSuchElementException::class)
    fun getOrCreateChat(identifier: ChatIdentifier): Chat

    fun getOrCreateChat(
        owner: UUID,
        partner: UUID
    ) = getOrCreateChat(toIdentifier(owner, partner))

    fun listChat(
        owner: User,
        version: Int,
        offset: Chat,
        limit: Int = CONVERSATION_LIMIT
    ): List<Chat>

    fun listChat(
        owner: User,
        version: Int,
        offset: ChatIdentifier,
        limit: Int = CONVERSATION_LIMIT
    ): List<Chat>

    fun listChat(
        owner: User,
        version: Int
    ): List<Chat>

    fun listChat(owner: User): List<Chat>

    fun getSnapshot(
        chat: Chat,
        owner: User,
        atVersion: Int
    ): ChatSnapshot
}
